---
title: "Gene Expression"
author: "Justina Kraujūnaitė"
date: "7/28/2021"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

In order to answer the big question of this project - **What is the difference between gene expression in fetal and adult brains?** - I will re-analyse the data of Jaffe et al. study (http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4281298/). In this research, the authors were looking for genes that showed patterns of expression that changed over time as people aged. Authors collected 48 different samples and studied 6 different age groups from fetal (<0 years) to old (50+ years). 

Since the samples themselves are quite big, in this work I will collect down to only 3 samples for each fetal (<0 years) and adult (20-50 years) group.

# Obtaining the raw data

I used Galaxy (https://usegalaxy.org) and downloaded all 6 datasets from the European Nucleotide Archive (ENA) (https://www.ebi.ac.uk/ena/browser/home). Fetal samples' runs' ids: SRR1554537, SRR1554566, SRR1554568. Adult samples' runs' ids: SRR1554536, SRR1554539, SRR1554534. I chose these specific samples regarding to their small size in comparison to others. Each samples' file was paired-end library and there were two fastq.gz format files for each sample.

Also, I collected the meta-data (phenotypic information and technological information) for each sample from BioSample database (https://www.ncbi.nlm.nih.gov/biosample/) and stored it in a tab-delimited text file (*phenotype_table.txt*), where each column corresponds to a variable and each row corresponds to one sample's value for those variables. I assembled information on the age (Age) and sex (Sex) of the person who the brain came from and the RNA-quality (RIN) of the sample. The age is measured as weeks after conception converted to years before term, which is why the variable Age is negative for some samples (it indicates the fetal samples).

# Aligning the data to the genome

To pre-process the data I used HISAT2 (Galaxy Version 2.1.0+galaxy7) and aligned downloaded samples to its' reference genome: built-in human genome version hg19; paired-end library with two "fastqsanger" files for each sample. 

The output was 6 BAM files. The number of aligned reads for each sample: SRR1554537 - 54 659 776; SRR1554566 - 52 773 411; SRR1554568 - 47 788 193; SRR1554536 - 21 308 422; SRR1554539 - 33 239 440; SRR1554534 - 27 765 625.

# Quality control on the alignments

I used FastQC (Galaxy Version 0.72+galaxy1) to check the quality of the alignments. I stored the QC results for each sample next to meta-data I collected earlier. It showed that all data are in a good shape, meaning appropriate alignments rates (>99%) and quality scores of mapped reads (average 37-38).

```{r}
phenotype_table <- read.delim("phenotype_table.txt")
phenotype_table
```

```{r}
fetal_samples <- phenotype_table[1:3,]
adult_samples <- phenotype_table[4:6,]

summary(fetal_samples[6])
summary(adult_samples[6])
```

The mapping rates are similar for fetal (mean: 99.74%) and adult (mean: 99.76%) samples.

Here is an example of full FastQC Report on SRR1554537 sample: https://usegalaxy.org/datasets/bbd44e69cb8906b51c6e30f85bda44d8/display?to_ext=html.

# Expression measurements at the gene count level 

To calculate the abundance of every gene in every sample, I used featureCounts (Galaxy Version 2.0.1). Gene annotation file was built-in genome hg19. The results were one table for each sample that was formatted with one gene per row and corresponding read-count. Then I merged all 6 tables into one table that was formatted with one gene per row and one sample per column.

```{r, message = FALSE, warning = FALSE}
library(org.Hs.eg.db)
library(annotate)
```


```{r}
# Reading feature count files

list_files <- list.files(path = "/Users/jkcollection/Downloads/Capstone/FeatureCounts/", pattern = "*.tabular", full.names = TRUE)
read_files <- lapply(list_files, read.delim)

# Merging files by gene id

gene_expression_table <- Reduce(function(f1, f2) merge(f1, f2, by = "Geneid"), read_files)
colnames(gene_expression_table) <- c("GeneID", "SRR1554537", "SRR1554566", "SRR1554568", "SRR1554536", "SRR1554539", "SRR1554534")

# Converting gene id to gene name

for (i in 1:nrow(gene_expression_table)){
    gene_expression_table[i,1] = lookUp(toString(gene_expression_table[i,1]), "org.Hs.eg", "SYMBOL")
}
rownames(gene_expression_table) <- make.names(gene_expression_table[,1], unique = TRUE)
gene_expression_table[,1] <- NULL
head(gene_expression_table)

# Writing table to a file

write.table(gene_expression_table, file = "gene_expression_table.txt", sep = "\t", row.names = TRUE, col.names = TRUE)
```

Here is a Galaxy workflow for all 3 fetal samples (the same was done with 3 adult samples):

![](/Users/jkcollection/Downloads/Capstone/Galaxy_Workflow.png)

# Exploratory analysis

Now I will explore the data for important features.

```{r, message = FALSE, warning = FALSE}
library(SummarizedExperiment)
library(edgeR)
```

```{r}
# Removing low expression data

gene_expression_table <- gene_expression_table[rowMeans(gene_expression_table) > 10,]

# Creating SummarizedExperiment object

SummExp <- SummarizedExperiment(assays = list(counts = gene_expression_table), colData = phenotype_table, rowData = rownames(gene_expression_table))
SummExp

# Making a boxplot of the expression levels for each sample 

dge_list <- DGEList(counts = assay(SummExp, "counts"), group = phenotype_table$Age_group)
dge_list$samples <- merge(dge_list$samples, as.data.frame(colData(SummExp)), by = 0)
boxplot(dge_list$counts)
```

Since the data is skewed, I will apply log2 transformation to have a clear view.

```{r}
log2_dge_list_counts <- log2(dge_list$counts + 1)
boxplot(log2_dge_list_counts)
```

Now boxplots are clearly visible. It seems that there are many outliers with extremely high gene expression in all adult samples, but not in fetal samples.

Next I will perform a principal components analysis (PCA) on the data and plot the top two principal components.

```{r, message = FALSE, warning = FALSE}
library(ggfortify)
```

```{r}
# Performing PCA

pca_counts <- prcomp(log2_dge_list_counts, center = TRUE, scale. = TRUE)
pca_counts

# Plotting PC1 and PC2, coloring by RIN, shaping by Age_group

PC <- data.frame(PC1 = pca_counts$rotation[,1], PC2 = pca_counts$rotation[,2], Age_group = phenotype_table$Age_group, RIN = phenotype_table$RIN)
ggplot(PC, aes(x = PC1, y = PC2, color = RIN, shape = Age_group)) + geom_point(size = 5)
```

From the scatter plot, it's easy to differentiate fetal samples from adult samples by PC1 and PC2, but it's hard to do it only by RIN.

# Statistical analysis

Next I will perform a statistical analysis to detect genes that are differentially expressed between fetal brains and adult brains.

H0 - the genes expression level is equal between fetal and adult brains.  
H1 - the genes expression level is different between fetal and adult brains.

To test the null and alternative hypotheses, I will fit a linear model.

```{r, message = FALSE, warning = FALSE}
library(limma)
library(edge)
```

```{r, message = FALSE}
# Making log2 transformation and removing low expression data

expression_data <- assay(SummExp)
expression_data <- log2(as.matrix(expression_data) + 1)
expression_data <- expression_data[rowMeans(expression_data) > 10,]

# Fitting linear model by Age_group

model_matrix <- model.matrix(~ SummExp$Age_group)
linear_model <- lmFit(expression_data, model_matrix)
ebayes_model <- eBayes(linear_model)
linear_model_toptable <- topTable(ebayes_model, number = dim(expression_data)[1])
```

```{r}
# Outputting only gene name, log2 fold-change estimate, p-value for that gene and adjusted p-value

linear_model_toptable_output <- linear_model_toptable[,c(1, 4, 5)]
head(linear_model_toptable_output)

# Writing table to a file

write.table(linear_model_toptable_output, file = "diff_exp_genes.txt", sep = "\t", row.names = TRUE, col.names = TRUE)
```

Now I will use a correction to identify genes that are differentially expressed after accounting for multiple testing. I will make a plot (volcano plot) of the fold-change for age in each linear model versus the log10 p-value.

```{r}
# Making a volcano plot and marking genes with p-value < 0.05 as blue

with(linear_model_toptable, plot(logFC, -log10(adj.P.Val)))
with(subset(linear_model_toptable, adj.P.Val < 0.05), points(logFC, -log10(adj.P.Val), col = "blue"))
```

Genes in blue are considered as differentially expressed.

Next I will summarise how many genes are differentially expressed at a given error rate, which is adjusted p-value < 0.05.

```{r}
print(sum(linear_model_toptable_output$adj.P.Val < 0.05))
print(sum(linear_model_toptable_output$adj.P.Val < 0.05 & linear_model_toptable_output$logFC > 0))
print(sum(linear_model_toptable_output$adj.P.Val < 0.05 & linear_model_toptable_output$logFC < 0))
```

3 864 genes are differentially expressed between fetal and adult brains. 3 693 genes are up-regulated and 171 genes are down-regulated from fetal to adult.

# Gene set analysis

Now I will examine differentially expressed genes in a wider context and try to identify biological patterns while answering these questions:  
Are there changes in H3K4me3 between fetal and adult brain over promoters for genes differentially expressed between fetal and adult brain?  
Are promoters of genes differentially expressed between adult and fetal brain marked by H3K4me3 in liver?

```{r, message = FALSE, warning = FALSE}
library(AnnotationHub)
```

```{r, message = FALSE, warning = FALSE}
# Downloading fetal brain, adult brain and adult liver data for the promoter associated histone modification H3K4me3 from roadmap epigenomics project

ah <- AnnotationHub()
ah_HS <- subset(ah, species == "Homo sapiens")
ah_fetal <- query(ah_HS, c("EpigenomeRoadMap", "E081", "H3K4me3"))
ah_adult <- query(ah_HS, c("EpigenomeRoadMap", "E073", "H3K4me3"))
ah_liver <- query(ah_HS, c("EpigenomeRoadMap", "E066", "H3K4me3"))

# Downloading narrowPeak datasets

np_fetal <- ah_fetal[[2]]
np_adult <- ah_adult[[2]]
np_liver <- ah_liver[[2]]
```

```{r, message = FALSE, warning = FALSE}
library(mygene)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

```{r, message = FALSE, warning = FALSE, results = 'hide'}
# Converting genes names to entrez genes ids

diff_exp_genes <- rownames(linear_model_toptable[linear_model_toptable$adj.P.Val < 0.05,])
diff_exp_genes_ids <- queryMany(diff_exp_genes, scopes = "symbol", fields = "entrezgene", species = "human")

# Extracting promoters of differentially expressed genes

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
txdb_genes <- genes(txdb)
diff_exp_promoters <- promoters(txdb_genes[diff_exp_genes_ids$entrezgene %in% txdb_genes$gene_id])
```

```{r}
# Finding the overlap between promoters of differentially expressed genes and narrowPeak datasets

fetal_overlap_percentage <- length(subsetByOverlaps(np_fetal, diff_exp_promoters, ignore.strand = TRUE)) / length(np_fetal) * 100
round(fetal_overlap_percentage, 1)
adult_overlap_percentage <- length(subsetByOverlaps(np_adult, diff_exp_promoters, ignore.strand = TRUE)) / length(np_adult) * 100
round(adult_overlap_percentage, 1)
```

36.8% promoters of differentially expressed genes overlap with fetal brain narrowPeak data, 24.5% - with adult brain narrowPeak data. This means that there are changes in H3K4me3 between fetal and adult brain over promoters for genes differentially expressed between fetal and adult brain.

```{r}
liver_overlap_percentage <- length(subsetByOverlaps(np_liver, diff_exp_promoters, ignore.strand = TRUE)) / length(np_liver) * 100
round(liver_overlap_percentage, 1)
```

20.3% promoters of differentially expressed genes overlap with liver narrowPeak data. This indicates that less promoters of genes differentially expressed between fetal and adult brain are marked by H3K4me3 in liver.

# Session info

```{r}
sessionInfo()
```
